### PR TITLE
Suppress object transfer requests when object is already being received.

### PR DESCRIPTION
**Pull request by elibol**
## What do these changes do?

**TL;DR: Corrects an edge case where an object is requested after it has been received.**

When a 1GB object is being received in 4096 chunks, the last iteration currently takes about 335ms. During this time, the `fetch_timeout_handler` timeout may have elapsed. We expect `process_add_object_notification` to be invoked in subsequent event loops so that the received `object_id` is inserted into `local_available_objects`. This results in a race condition between `process_add_object_notification` and `fetch_timeout_handler`. If `fetch_timeout_handler` is invoked before `process_add_object_notification`, then `fetch_timeout_handler` will issue another request for the object. The object will be resent, and subsequently handled by `ignore_read_chunk` since `PlasmaClient` will fail to create the object for a read (by this time, the mechanism which detects the object's presence presumably reports the object as being present). The event loop invokes these handlers with no known guaranteed ordering. If event handler invocation order is non-deterministic, then local object detection may sometimes fail. If the event loop invokes handlers in chronological order (e.g. if `fetch_timeout_handler` timeout elapsed before `process_add_object_notification` was received, then `fetch_timeout_handler` would be invoked before `process_add_object_notification`), then the probability of `fetch_timeout_handler` being called before `process_add_object_notification` increases with the size of the object being transferred.

The time it takes to compute the last iteration is proportional to the size of the object being transferred (times are in microseconds):
obj_size,chunk_size,mean_rcv_us,**last_rcv_us**,last_snd_us
1gb,4096,8,**335566**,36343
100mb,4096,7,**11492**,6376
1mb,4096,12,**281**,5
1kb,4096,27,**27**,30
1b,4096,34,**34**,49

The above hypothesis was tested by adding a `received_objects` map to `PlasmaManagerState`, which is set immediately after `Seal` and `Release` are invoked. This map is checked for the existence of an `object_id` whenever `fetch_timeout_handler` is deciding which `object_ids` to request. As soon as `process_add_object_notification` is invoked, the `object_id` is removed from `received_objects` and added to `local_available_objects` instead (to preserve existing functionality and not grow `received_objects` indefinitely). The issue no longer occurs with these changes in place.

## Related issue number

None

### CODE REVIEW

**Comment by robertnishihara** [comment]
Can you say which concrete benchmark gets faster after this change? And how much faster does it get?

File: src/plasma/plasma_manager.cc [hunk 1]
```diff
@@ -940,6 +948,20 @@ int fetch_timeout_handler(event_loop *loop, timer_id id, void *context) {
        it != manager_state->fetch_requests.end(); it++) {
     FetchRequest *fetch_req = it->second;
     if (fetch_req->manager_vector.size() > 0) {
+      if(is_object_received(manager_state, fetch_req->object_id)){
+        int64_t duration = current_time_ms()
+                           - manager_state->received_objects[fetch_req->object_id];
+        if(duration > 10 * RayConfig::instance().manager_timeout_milliseconds()){
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: I'd prefer not to hardcode this value `10` here, maybe add it to `ray_config.h`? What considerations should affect the value of this constant?

File: src/plasma/plasma_manager.cc [hunk 2]
```diff
@@ -692,6 +699,7 @@ void process_data_chunk(event_loop *loop,
     ARROW_CHECK_OK(plasma_conn->Release(buf->object_id.to_plasma_id()));
     /* Remove the request buffer used for reading this object's data. */
     conn->transfer_queue.pop_front();
+    conn->manager_state->received_objects[buf->object_id] = current_time_ms();
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Can you add a comment above like

```
// Record the time at which the manager finished receiving this object.
```

File: src/plasma/plasma_manager.cc [hunk 3]
```diff
@@ -235,6 +235,11 @@ struct PlasmaManagerState {
   /** The time (in milliseconds since the Unix epoch) when the most recent
    *  heartbeat was sent. */
   int64_t previous_heartbeat_time;
+  /**
+   * Objects are mapped to the time, in milliseconds, they were received. Objects
+   * are removed from this map after they are added to local_available_objects.
+   */
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Can you elaborate a bit? When precisely are the objects added to this map? Is it when the transfer starts or the transfer finishes? What is the "time they were received"? Is it when the transfer started or completed?

File: src/plasma/plasma_manager.cc [hunk 4]
```diff
@@ -235,6 +235,11 @@ struct PlasmaManagerState {
...
+  std::unordered_map<ObjectID, int64_t, UniqueIDHasher> received_objects;
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: If I understand correctly, this map contains precisely the objects where the transfer has been completed but the store hasn't sent a notification to the manager yet. Is that right?

**Review by stephanie-wang — COMMENTED** [review]
If I understand this correctly, are you tracking the time since you sealed an object to when its next fetch request would go out, to decide whether it's been long enough that maybe the object addition to the store actually failed? I'm not sure if this is necessary since I believe the plasma manager will fail if the store that it subscribes to fails, so the manager will eventually get notified of any seals if there are no failures. It may be better to just make the mapping a set of object IDs with no timestamps.

Also, I wonder if it would be nicer to change the semantics of the map to "all objects that we have received some bytes of, but for which we haven't received an object notification from the store". At the moment, we deduplicate requests of the same object on the plasma manager that serves the request, but this would also let us deduplicate on the requesting manager. We would just have to be careful here to remove objects from `received_objects` if the manager that was serving the transfer dies (in that case I think the race condition you describe could still occur, but it should be uncommon).

File: src/plasma/plasma_manager.cc (hunk 3 above)
Comment (elibol) [reply src/plasma/plasma_manager.cc]: Will clarify in documentation.

File: src/plasma/plasma_manager.cc (hunk 4 above)
Comment (elibol) [reply src/plasma/plasma_manager.cc]: yes.

**Comment by elibol** [comment]
Yes @stephanie-wang, you've understood it correctly. Thanks, this is great feedback. Just to be clear, if the plasma store fails to broadcast the event that triggers `process_add_object_notification`, then the plasma store will exit/fail, and the plasma manager is designed to fail if the plasma store fails. If this is correct, I'll remove that block of code and convert `received_objects` to an `unordered_set`.

I like the idea of changing the semantics. Currently, how does the receiving manager handle the serving manager's death?

File: src/plasma/plasma_manager.cc (hunk 1 above)
Comment (elibol) [reply src/plasma/plasma_manager.cc]: The primary consideration is the time interval `delta` after the plasma manager receives an object and before the plasma manager receives the message which triggers  `process_add_object_notification` for the object received from the plasma store. `10 * RayConfig::instance().manager_timeout_milliseconds()` can be thought of as an upper bound on `delta`, after which the plasma manager assumes the receive has failed. I don't think this is necessary given @stephanie-wang's feedback.

**Comment by stephanie-wang** [comment]
Yes, @elibol, that's right, the plasma manager will also fail ([code](https://github.com/ray-project/ray/blob/master/src/plasma/plasma_manager.cc#L1329)).

Right now, the next notification that the receiving manager gets on the socket for transferring data will be an error. The receiving manager notifies the store to abort the object and then cleans up the connection state with the sending manager. [This](https://github.com/ray-project/ray/blob/master/src/plasma/plasma_manager.cc#L677) is the relevant line.

**Comment by robertnishihara** [comment]
@stephanie-wang the plasma manager dies if the corresponding *store* dies. However, I think @elibol was asking about the case where manager 1 is receiving an object from manager 2 and manager 2 dies. That case is handled by this code I believe

https://github.com/ray-project/ray/blob/0b022c097391bccc21cb07a20933a3b85a053a97/src/plasma/plasma_manager.cc#L676-L683

File: src/plasma/plasma_manager.cc [hunk 5]
```diff
@@ -528,6 +536,11 @@ void PlasmaManagerState_free(PlasmaManagerState *state) {
   delete state;
 }
 
+bool is_object_received(PlasmaManagerState *state, ObjectID object_id){
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: `const ObjectID &object_id`

Comment (concretevitamin) [reply src/plasma/plasma_manager.cc]: If you const the 2nd arg, it leaves a hint that the 1st arg is an out-arg, which doesn't seem to be the case.  So please const both :)

File: src/plasma/plasma_manager.cc [hunk 6]
```diff
@@ -832,6 +843,16 @@ void process_data_request(event_loop *loop,
...
                           ClientConnection *conn) {
+
+  /** Monitor objects that are in progress of being received.
+   *  If a read fails while receiving this object, its
+   *  ObjectID will be removed. If the object is successfully
+   *  received, its ObjectID is removed by process_add_object_notification.
+   *  If a shared buffer for the object cannot be created,
+   *  then the receive is ignored, and the corresponding ObjectID
+   *  is removed from receives_in_progress in this function. */
+  conn->manager_state->receives_in_progress.insert(object_id);
```
Comment (stephanie-wang) [inline src/plasma/plasma_manager.cc]: Maybe move this line to the if-else block at the end of this function so that we don't have to erase?

File: src/plasma/plasma_manager.cc [hunk 7]
```diff
@@ -1031,7 +1065,11 @@ void object_table_subscribe_callback(ObjectID object_id,
       db_client_table_get_ip_addresses(manager_state->db, manager_ids);
   /* Run the callback for fetch requests if there is a fetch request. */
   auto it = manager_state->fetch_requests.find(object_id);
-  if (it != manager_state->fetch_requests.end()) {
+
+  if (it != manager_state->fetch_requests.end() &&
+      !is_receiving_or_received(manager_state, object_id)) {
```
Comment (stephanie-wang) [inline src/plasma/plasma_manager.cc]: I would move the `is_receiving_or_received` check inside `request_transfer` when deciding whether to call `request_transfer_from`, so that we still record the manager locations received. Perhaps we should also rethink the naming of `request_transfer`...

**Review by stephanie-wang — COMMENTED** [review]
Nice! This is pretty clean :)

**Review by stephanie-wang — APPROVED** [review]
I wish that we had a good way to test this, but this looks good assuming Travis passes!

File: src/plasma/plasma_manager.cc [hunk 8]
```diff
@@ -682,6 +682,9 @@ void process_data_chunk(event_loop *loop,
   int err = read_object_chunk(conn, buf);
   auto plasma_conn = conn->manager_state->plasma_conn;
   if (err != 0) {
+    /** Remove the object from the receives_in_progress set so that
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: This comment style should only be used for structs/functions where the comment is intended to be parsed by doxygen to autogenerate documentation. Instead, use

```
/*
 * ... */
```

or

```
//
//
```

we're switching to the latter, but either is ok for now.

File: src/plasma/plasma_manager.cc [hunk 9]
```diff
@@ -949,9 +966,9 @@ int fetch_timeout_handler(event_loop *loop, timer_id id, void *context) {
        it != manager_state->fetch_requests.end(); it++) {
     FetchRequest *fetch_req = it->second;
     if (fetch_req->manager_vector.size() > 0) {
-      if(is_object_received(manager_state, fetch_req->object_id)) {
-        // Do nothing if the object has already been received.
-        LOG_DEBUG("fetch_timeout_handler: Object received. %s",
+      if (is_receiving_or_received(manager_state, fetch_req->object_id)) {
+        // Do nothing if the object is in progress or has already been received.
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: "object is in progress" -> "object transfer is in progress"

File: src/plasma/plasma_manager.cc (hunk 8 above)
Comment (robertnishihara) [reply src/plasma/plasma_manager.cc]: This applies to other places as well.

File: src/plasma/plasma_manager.cc [hunk 10]
```diff
@@ -235,12 +235,11 @@ struct PlasmaManagerState {
   /** The time (in milliseconds since the Unix epoch) when the most recent
    *  heartbeat was sent. */
   int64_t previous_heartbeat_time;
-  /** An object's object_id is added to this set immediately after
-   *  it is received and sealed. The object is removed in
-   *  process_add_object_notification, which is triggered by
-   *  the corresponding notification from the plasma store. */
-  // TODO(hme): change semantics to "received some bytes of object"
-  std::unordered_set<ObjectID, UniqueIDHasher> received_objects;
+  /** An ObjectID is added to this set if a shared buffer is
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Can you add something like "This is the subset of the ObjectIDs currently being transferred to this manager" As well as "Note that there is no guarantee that an ObjectID will appear in this set even if it is currently being transferred to this manager (because if a single transfer fails, we flush the entire set)"

File: src/plasma/plasma_manager.cc [hunk 11]
```diff
@@ -235,12 +235,11 @@ struct PlasmaManagerState {
...
+   *  successfully created for the corresponding object.
+   *  The ObjectID is removed in process_add_object_notification, which is
+   *  triggered by the corresponding notification from the plasma store. */
+  std::unordered_set<ObjectID, UniqueIDHasher> receives_in_progress;
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Are we handling the case where a remote manager dies? In that case, we should flush the entire set.

Comment (stephanie-wang) [reply src/plasma/plasma_manager.cc]: @elibol erases the entry when the object is aborted. Assuming we keep the current design where we only receive one object at a time from each remote manager, I don't think it's necessary to flush the entire set.

Comment (robertnishihara) [reply src/plasma/plasma_manager.cc]: I see, I missed that. I was thinking that we might implement something like #1485 which would introduce a bug. It wouldn't cause any test failures and it would only happen rarely.

I'm ok with either case, but if we leave it the way it is we should document very clearly the assumption being made about only one object being transferred at a time from a given manager.

I think flushing the whole set is ok because 1) the set should always be pretty small anyway and 2) reissuing a transfer request accidentally is not the end of the world.

File: src/plasma/plasma_manager.cc [hunk 12]
```diff
@@ -674,6 +687,9 @@ void process_data_chunk(event_loop *loop,
   int err = read_object_chunk(conn, buf);
   auto plasma_conn = conn->manager_state->plasma_conn;
   if (err != 0) {
+    // Remove the object from the receives_in_progress set so that
+    // retries are processed.
+    conn->manager_state->receives_in_progress.erase(buf->object_id);
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: Could you add an additional comment here with a TODO about flushing the whole thing if we allow parallel object transfers? I saw the comment above, but it would be good to include here as well.

File: src/plasma/plasma_manager.cc [hunk 13]
```diff
@@ -1296,6 +1336,11 @@ void process_add_object_notification(PlasmaManagerState *state,
                                      int64_t metadata_size,
                                      unsigned char *digest) {
   state->local_available_objects.insert(object_id);
+  if (state->receives_in_progress.count(object_id) > 0) {
+    /** This object is now locally available, so remove it from the
+     *  receives_in_progress set. */
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: comment style here

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
Remaining tests passed on ray-private-travis
